### ISSUE TITLE
DataAnnotationLocalizerProvider documentation wrong/incomplete

### ISSUE THREAD

**Issue by WolfspiritM** [aspnet/Mvc#7437]
Hi,

I'm trying to localize DataAnnotations like "[Require]".
I followed: https://docs.microsoft.com/en-us/aspnet/core/fundamentals/localization#using-one-resource-string-for-multiple-classes

I created a Resource called "DataAnnotationsResource.resx" in the "Resources" folder and added the following to my Startup.cs:

```cs
            services.AddLocalization(options => options.ResourcesPath = "Resources");

            services.AddMvc()
                .AddViewLocalization(LanguageViewLocationExpanderFormat.Suffix)
                .AddDataAnnotationsLocalization(options =>
                {
                    options.DataAnnotationLocalizerProvider = (type, factory) =>
                       factory.Create(typeof(DataAnnotationResource));
                });
```

The Localization of the whole page works however DataAnnotations are not translated. If I remove DataAnnotationLocalizerProvider and specify the translation inside the model.resx it works.

Then I tested a bit and created a Controller which injects IStringLocalizerFactory and I tried to do the same as the function above:
```cs
            var createdlocalizer = factory.Create(typeof(DataAnnotationResource));
            return Ok(createdlocalizer["Confirm password"]);
```

And that results in:
```json
{"name":"Confirm password",
"value":"Confirm password",
"resourceNotFound":true,
"searchedLocation":"MyWebApp.Resources.Resources.DataAnnotationResource"}
```

So it adds "Resources." to the 'Path'.
I think that is a correct behaviour for Controller and View Localization where the Factory gets called with the Type of the Controller and gets the corresponding Resource from the Resource folder but in the case that is explained by the Docs it breaks Localization for DA as you can not Create an Instance of IStringLocalizer directly from the Resource Class.

A workaround would be to create a dummy "DataAnnotationResource" class outside of the Resource folder which seems a bit ugly.

Is there any better way to do that?

**Comment by mkArtakMSFT** [comment]
Hi @WolfspiritM. Thanks for reaching out to us. @ryanbrandenburg, do you have any thoughts or suggestions here?

**Comment by ryanbrandenburg** [comment]
When I attempted to reproduce your scenario it worked as expected. My first guess at what trouble you might be having is that you created a resource at "DataAnnotationsResource.resx" but the type you're trying to look it up with is "DataAnnotationResource" (no 's'). If that was just a typo in copying things to the issue or you're still having trouble the next thing I'll need to help you diagnose this is a git repo which reproduces the problem.

**Comment by WolfspiritM** [comment]
@ryanbrandenburg Thanks. It's not a typo (it is a typo in my post above however as the file is called DataAnnotationResource.resx). I created a sample MC Repo here: https://github.com/WolfspiritM/LocalizationIssue

Once started on the Index Page it will Display an input box and a label for the IndexViewModel Property "Username".
The Label is "Username" even so I defined a different text in the Resource.
Also I've set a StringLength on the Property and if you enter a single character it displays "LengthNotMatch" instead of the text I defined inside the Resource.
Also I added a controller that is reachable via .../translation which returns the Result from the StringLocalizer (from DI and from a factory (which is both the same)) that is accessing the Resource.
The searchLocation says: "LocalizationIssue.Resources.Resources.DataAnnotationResource" which I think is the problem.

It would be nice to have a "check" in there if the requested Type is already within "Resource" so it doesn't use the relative Path with added "ResourcesPath".

Also I created a branch "workaround" which adds a dummy class to the root of the project just for resolution of the correct Resource. In that branch it works as expected but it feels "ugly".

**Comment by ryanbrandenburg** [comment]
You're correct that getting "LocalizationIssue.Resources.Resources.DataAnnotationResource" is your problem. That happens because the `Designer.cs` that VS created for you is in the namespace `LocalizationIssue.Resource`. If you change the namespace for that to be `LocalizationIssue` it should work. Further, those Designer files aren't super relevant to the way AspNetCore MVC does localization, so  you don't really need one. As you noted another option would be creating a "dumby" class, but if you don't like doing that you can keep your "base" localizations in a `Startup.resx` and use the startup class to access it.

Such checks about types already being within "Resources" are dangerous, it's perfectly legal for people to have resources under "LocalizationIssue.Resources.Resources.DataAnnotationResource" if they wanted, or to have ResourcesPath values other than "Resources" for which having the repeated value would make even more sense.

**Comment by WolfspiritM** [comment]
Thanks for the response.

If I remove the Designer file completly the compiler doesn't know anything about the "DataAnnotationResource" type at all resulting in:
```
Startup.cs(37,51): error CS0246: The type or namespace name 'DataAnnotationResource' could not be found (are you missing a using directive or an assembly reference?)
```
So at least for compile time they are relevant if I want to use typeof(DataAnnotationResource) as proposed in the docs. Changing the Namespace of the Designer file manually works but is that the intended way this should be done?

Using Startup is a good idea however I have the same problem for a SharedLocalizer and I'd like to seperate these. I ended up creating dummy classes for that aswell.

I was looking into it a bit more and found this:
https://github.com/aspnet/Localization/blob/2.0.1/src/Microsoft.Extensions.Localization/ResourceManagerStringLocalizerFactory.cs#L20
It states the order is Attribute -> ResourcePath -> project root
If that is correct and `LocalizationIssue.Resources.Resources.DataAnnotationResource` (with added ResourcePath) was not able to find the Resource, shouldn't it then move on and search from the project root for `LocalizationIssue.Resources.DataAnnotationResource`?
That would actually fix this problem aswell.

If that is not the case and ResourcePath is forced even if no Resource is found at that location then:

As far as I understand there are two different ways someone might want to access the IStringLocalizerFactory:
1. The normal way by calling it with the actual type of the controller/view/... which is outside of the Resource folder/namespace and then gets mapped to the Resource file by appending LocalizationOptions.ResourcesPath to the Base Namespace. Controller.TranslationController -> Resources.Controller.TranslationController
2. Accessing the Resource directly from the Resource Namespace which is what the docs propose in this case. (I'm not sure if that is actually a supported way to use it)

I agree that hardcoding "Resources" is dangerous however the ResourcePath is defined dynamicly via AddLocalization. So what I meant with type check is: if the Factory gets called with a type that has a Namespace which lives within the "BasePath.ResourcePath" already I think it's guaranteed that the resulting StringLocalizer should use that Resource and not add ResourcePath to the Namespace again.

E.g. with ResourcePath = "Resources":
```factory.Create(typeof(App.Controllers.HomeController)``` (1. way to call create)
  -> Current: App.Resources.Controllers.HomeController.resx
  -> Expected: App.Resources.Controllers.HomeController.resx
```factory.Create(typeof(App.DataAnnotations)``` (1. way to call create)
  -> Current: App.Resources.DataAnnotations.resx
  -> Expected: App.Resources.DataAnnotations.resx
```factory.Create(typeof(App.Resources.DataAnnotations)``` (2. way to call create)
  -> Current: App.**Resources.Resources**.DataAnnotations.resx
  -> Expected: App.Resources.DataAnnotations.resx
```factory.Create(typeof(App.Resources.Resources.Controllers.HomeController)``` (2. way to call create)
  -> Current: App.Resources.Resources.Resources.Controllers.HomeController.resx
  -> Expected: App.Resources.Resources.Controllers.HomeController.resx

So while it is totally valid to have Resources within Resources I can't think of any case where you want to map a Resource Designer Class that resides within the `ResourcesPath` to another `ResourcesPath` beneath it. 
But maybe I'm just missing a case where that might not be true ?? 

Right now if you have a Resource Folder within the Resource Folder and have two Resources with the same name it might end up using the one you might not expect:

* Resources
  * DataAnnotationResource.resx
  * Resources
    * DataAnnotationResource.resx

```factory.Create(typeof(App.Resources.DataAnnotationResource))``` would actually use the ```App.Resources.Resources.DataAnnotationResource```

**Comment by ryanbrandenburg** [comment]
Yes, if you're using the `typeof()` overload of Create the type must exist. You're welcome to have a "dumby class", reference a related class (and change the name of your resx to reflect it), etc., so long as the results meet [these](https://docs.microsoft.com/en-us/aspnet/core/fundamentals/localization#resource-file-naming) criteria. If you don't want to change or remove the `Designer.cs` file, another option would be to just rename your resource to match it (ie `Resources\Resources.DataAnnotationResource.resx`).

The code comment you linked is perhaps a bit unclear. It means to convey that there are 3 ways to set the resource relative path, and that they are chosen if the resource path is provided via that means, not that it will check all three locations for resources. That is, if you have both `ResourceLocationAttribute` and `ResourcesPath` set, ONLY the `ResourceLocationAttribute` location will be checked, and if it doesn't exist you'll get `"resourceNotFound":true,`. It is not "fallback" but a selection order.

It's not clear to me what distinction you're attempting to draw when talking about the ways to use `IStringLocalizerFactory`. In your second use are you referring to [this](https://github.com/aspnet/Localization/blob/2.0.1/src/Microsoft.Extensions.Localization/ResourceManagerStringLocalizerFactory.cs#L159) overload? The examples you listed are all the same, they create a `StringLocalizer` from a type. That types full name is used to determine the resource to look for as described [here](https://docs.microsoft.com/en-us/aspnet/core/fundamentals/localization#resource-file-naming). The presence or absence of the `ResourcesPath` in the namespace is up to the creator to match on both ends as they see fit.

Does that answer your questions?

**Comment by mkArtakMSFT** [comment]
TL;DR, @ryanbrandenburg, are there any other actions here, should we close already?
